httpserver/swarm.cpp

```diff
@@ -123,23 +124,40 @@ SwarmEvents Swarm::update_swarms(const all_swarms_t& swarms) {
 swarm_id_t get_swarm_by_pk(const std::vector<SwarmInfo>& all_swarms,
                            const std::string& pk) {
 
-    // TODO: handle errors
-    // TODO: get rid of allocations?
+    if (pk.size() != 66) {
+        throw std::invalid_argument("invalid pub key size");
+    }
+
+    /// Create a buffer for 64 characters plus 4 nulls on 8-byte boundaries
+    char buf[68] = {};
+
+    /// Note: pk is expected to contain two leading characters
+    /// (05 for the messenger) that do not participate in mapping
+
+    memcpy(buf, pk.c_str() + 2, 16);
+    memcpy(buf + 17, pk.c_str() + 2 + 16, 16);
+    memcpy(buf + 34, pk.c_str() + 2 + 32, 16);
+    memcpy(buf + 51, pk.c_str() + 2 + 48, 16);
 
-    std::string pk0_str = std::string(pk.c_str() + 2, 16);
-    std::string pk1_str = std::string(pk.c_str() + 2 + 16, 16);
-    std::string pk2_str = std::string(pk.c_str() + 2 + 32, 16);
-    std::string pk3_str = std::string(pk.c_str() + 2 + 48, 16);
+    /// Note: if conversion is not possbile, we will still
+    /// get a value in res (possibly 0 or UINT64_MAX), which
+    /// we are not handling at the moment
+    uint64_t res = strtoull(buf, nullptr, 16);
+    res ^= strtoull(buf + 17, nullptr, 16);
+    res ^= strtoull(buf + 34, nullptr, 16);
+    res ^= strtoull(buf + 51, nullptr, 16);
```

Comment:
Here's another way to do this:
```cpp
#include <cstring>
...
  char buf[16 + 1]; // for null termination
  const char* src = pk.c_str() + 2; // skip the prefix
  uint64_t res = 0;
  for (int i = 0; i < pk.size() / 16; ++i)
  {
    strncpy(buf, src, 16);
    src += 16;
    res ^= strtoull(buf, nullptr, 16);
  }
```
Then you can remove `#include <stdlib.h>` as well.

The XOR operation won't exactly match what was there initially tho because of the initial `0 ^ x`, but it's good enough if the only purpose is to reduce the pubkey to a 64bit number.

Reply:
`0 ^ x` does not change anything in x, so it should produce the same result, correct?

Reply:
I rebased to fix the typo and compilation error (I used `rightmost` which should be `rightmost_id`). The reset is the same.

Reply:
Yup, brain fart. sorry!

Reply:
Updated to use a smaller buffer and combine values in a loop.

Reply:
Sweet